kong/runloop/handler.lua

```diff
@@ -58,16 +62,41 @@ end
 
 do
   -- Given a protocol, return the subsystem that handles it
-  local protocol_subsystem = {
-    http = "http",
-    https = "http",
-    tcp = "stream",
-    tls = "stream",
-  }
-
   local router
   local router_version
 
+  local function should_process_route(route)
+    for _, protocol in ipairs(route.protocols) do
+      if constants.PROTOCOLS_WITH_SUBSYSTEM[protocol] == subsystem then
+        return true
+      end
+    end
+
+    return false
+  end
+
+  local function get_service_for_route(db, route)
+    local service_pk = route.service
+    if not service_pk then
+      return nil, nil, subsystem
+    end
+
+    local service, err = db.services:select(service_pk)
+    if not service then
+      return nil, "could not find service for route (" .. route.id .. "): " ..
+                  err
+    end
+
+    -- TODO: this should not be needed as the schema should check it already
+    local service_subsystem = constants.PROTOCOLS_WITH_SUBSYSTEM[service.protocol]
+    if service_subsystem ~= subsystem then
+      log(WARN, "service with protocol '", service.protocol,
+                "' cannot be used with '", subsystem, "' subsystem")
```

Comment:
Should we bail out here? `return` with no error, simply log the warning and ignore the Route. Otherwise we may end up in situations where a particular subsystem's Route needs to Route into another subsystem (http to stream or vice-versa), which we are not ready to handle yet.

Or what's the rationale for keeping the Service attached to the Route, and do we have any tests covering this edge case (I assume not, since the schema should not even allow this, in which case, we are introducing untested/undefined behavior in our core logic).

Additionally, this way, we don't have to return 3 values from this function anymore, and callers of this function may use `subsystem` (the top-chunk upvalue) instead of `service_subsystem`.

Reply:
Yes, it was like that, but @james-callahan wanted it to be inverted and it introduced this bug. Damnit. I'll fix it.

Reply:
Made it return service OR nil, err OR nil (serviceless route).